src/libs/search.ts

```diff
@@ -0,0 +1,190 @@
+	searchItems: T[],
+	includeAllQuery = true
+): ScoreResultItem<T>[] => {
+	return searchItems.map((item: T) => {
+		let score = 0;
+		const matchedIndices: number[] = [];
+		for (const query of queries) {
+			const stringMenu = item.name;
+			let matchedIndex = 0;
+			let startIndex = 0;
+			let addedScore = 0;
+
+			// If found query
+			while ((matchedIndex = stringMenu.indexOf(query, startIndex)) >= 0) {
+				// Get the end index of matched query
+				const endMatchedIndex = matchedIndex + query.length;
+
+				// Add score by the length of matched query divided by the length of stringMenu
+				addedScore += query.length / stringMenu.length;
+
+				// Add matched indices
+				for (let i = matchedIndex; i < endMatchedIndex; i++) {
+					// Skip if already added
+					if (matchedIndices.includes(i)) continue;
+					matchedIndices.push(i);
+				}
+
+				// Set the start index of next search
+				startIndex = endMatchedIndex;
+			}
+
+			// If not found query
+			if (addedScore == 0) {
+				// If all query must be matched, return score 0
+				if (includeAllQuery) {
+					return {
+						item,
+						score: 0,
+						matchedIndices: []
+					};
+				} else {
+					// Penalty score if not all query is matched
+					addedScore -= 1 / queries.length;
+				}
+			}
+
+			// Add score
+			score += addedScore;
+		}
+
+		return {
+			item,
+			score,
+			matchedIndices
+		};
+	});
+};
+
+/**
+ * Calculates the score and highlights the matched indices of the top k candidates.
+ * @param candidates - The list of candidates to be scored and highlighted.
+ * @param k - The number of top candidates to return.
+ * @returns - The list of top candidates with highlighted matched indices.
+ */
+const postCalcuateScore = <T extends { name: string }>(
+	candidates: ScoreResultItem<T>[],
+	k = 5
+): ScoreAndHighlightResultItem<T>[] => {
+	return (
+		candidates
+			// Filter out score 0
+			.filter((c) => c.score > 0)
+			// Sort by score descending
+			.sort((a, b) => b.score - a.score)
+			// Take top k
+			.slice(0, k)
+			// Highlight matched indices
+			.map((c: ScoreResultItem<T>) => ({
+				...c,
+				highlightedName: highlightText(c.item.name, c.matchedIndices)
+			}))
+	);
+};
+
+/**
+ * Returns an array of objects representing the highlighted text.
+ * @param text - The text to highlight.
+ * @param indices - An array of indices representing the positions to highlight.
+ * @returns An array of objects representing the highlighted text.
+ */
+export function highlightText(text: string, indices: number[]): HighlightedText[] {
+	const result: HighlightedText[] = [];
+	for (let index = 0; index < text.length; index++) {
+		const char = text[index];
+		if (indices.includes(index) && !indices.includes(index - 1)) {
+			result.push({
+				text: '',
+				highlight: true
+			});
+		} else if (!indices.includes(index) && indices.includes(index - 1)) {
+			result.push({
+				text: '',
+				highlight: false
+			});
+		} else if (result.length === 0) {
```

Comment:
The result of this condition is the same as the above one. Better combine and use OR `||` to get rid of duplicated code.